File: calibration/calibration_SNOWPACK/dsnow_parameter_optimization.R
##############################################################################
# DELTASNOW PARAMETER OPTIMIZATION
##############################################################################
#
# Objective:
#   final_score = 0.5 * RMSE(SWE) + 0.5 * RMSE(bulk density)
#
# bulk density = SWE / HS
#
##############################################################################

library(optimx)
library(zoo)
library(foreach)
library(doParallel)
library(lubridate)
library(nixmass)
library(tidyverse)

##############################################################################
# SETTINGS
##############################################################################

season_start   <- "-08-01"
season_end     <- "-07-31"
start_of_block <- 8

##############################################################################
# LOAD DATA
##############################################################################

d_obs <- get(load(
  "/Users/jakobwerkgarner/code/mt_dsnow/calibration/calibration_SNOWPACK/data/d_obs_SNOWPACK.rda"
))

##############################################################################
# HELPER FUNCTIONS
##############################################################################

set_season <- function(date, start_month = 8) {
  date <- as.Date(date)
  yr <- year(date)
  mo <- month(date)
  ifelse(mo < start_month, yr - 1, yr)
}

rmse <- function(obs, mod) {
  ok <- is.finite(obs) & is.finite(mod)
  if (!any(ok)) return(NA_real_)
  sqrt(mean((mod[ok] - obs[ok])^2))
}

combined_rmse <- function(df, eps = 1e-6) {
  rmse_swe <- rmse(df$swe_obs, df$swe_mod)

  rho_obs <- ifelse(is.finite(df$hs) & df$hs > eps, df$swe_obs / df$hs, NA_real_)
  rho_mod <- ifelse(is.finite(df$hs) & df$hs > eps, df$swe_mod / df$hs, NA_real_)

  rmse_rho <- rmse(rho_obs, rho_mod)

  0.5 * rmse_swe + 0.5 * rmse_rho
}

##############################################################################
# SPLIT DATA INTO FIT AND VALIDATION
##############################################################################

d_obs_fit <- list()
d_obs_val <- list()

for (station in names(d_obs)) {

  cat("Processing station:", station, "\n")

  d <- d_obs[[station]]
  years <- unique(year(index(d)))

  fit_list <- list()
  val_list <- list()
  counter <- 1

  for (y in years[1:(length(years) - 1)]) {

    winter <- subset(
      d,
      index(d) >= as.Date(paste0(y, season_start)) &
        index(d) < as.Date(paste0(y + 1, season_end)) + 1
    )

    if (nrow(winter) < 200) {
      cat("  skipping year", y, "- only", nrow(winter), "values\n")
      next
    }

    if (
      nrow(winter) < 365 &&
      (as.numeric(winter$Hobs[1]) > 0.05 ||
       as.numeric(winter$Hobs[nrow(winter)]) > 0.05)
    ) {
      cat("  skipping year", y, "- incomplete winter with snow at edge\n")
      next
    }

    if (counter %% 2 == 0) {
      fit_list[[length(fit_list) + 1]] <- winter
    } else {
      val_list[[length(val_list) + 1]] <- winter
    }

    counter <- counter + 1
  }

  d_obs_fit[[station]] <- if (length(fit_list) > 0) do.call(rbind, fit_list) else NULL
  d_obs_val[[station]] <- if (length(val_list) > 0) do.call(rbind, val_list) else NULL
}

##############################################################################
# PREPARE FIT DATA
##############################################################################

fit_data_list <- list()

for (station in names(d_obs_fit)) {

  x <- d_obs_fit[[station]]

  if (is.null(x) || length(x) == 0) next

  dat <- tryCatch(as_tibble(coredata(x)), error = function(e) NULL)
  if (is.null(dat) || ncol(dat) == 0) next

  if (ncol(dat) == 2 && all(is.na(colnames(dat)))) {
    colnames(dat) <- c("Hobs", "SWEobs")
  }

  if (!all(c("Hobs", "SWEobs") %in% colnames(dat))) {
    cat("Skipping station", station, "- missing Hobs or SWEobs\n")
    next
  }

  fit_data_list[[station]] <- tibble(
    date    = as.Date(index(x)),
    name    = station,
    hs      = dat$Hobs,
    swe_obs = dat$SWEobs,
    block   = set_season(index(x), start_of_block)
  )
}

d_obs_fit_tibble <- bind_rows(fit_data_list)

##############################################################################
# OBJECTIVE FUNCTION
##############################################################################

minimize_score <- function(par, data, scale, verbose = FALSE) {

  par_real <- par * scale

  cat("pars =", paste(round(par_real, 6), collapse = ", "), "\n")

  station_results <- foreach(
    station = unique(data$name),
    .packages = c("dplyr", "tidyr", "nixmass")
  ) %dopar% {

    if (verbose) cat("station:", station, "\n")

    data_station <- data %>% filter(name == station)
    blocks <- unique(data_station$block)

    block_results <- lapply(blocks, function(b) {

      df_block <- data_station %>% filter(block == b)

      full_dates <- tibble(
        date = seq(min(df_block$date), max(df_block$date), by = "1 day")
      )

      joined <- df_block %>%
        select(date, hs, swe_obs) %>%
        right_join(full_dates, by = "date") %>%
        arrange(date)

      swe_mod <- tryCatch(
        {
          joined %>%
            select(date, hs) %>%
            mutate(date = as.character(date)) %>%
            nixmass::swe.delta.snow(
              model_opts = list(
                rho.max  = par_real[1],
                rho.null = par_real[2],
                c.ov     = par_real[3],
                k.ov     = par_real[4],
                k        = par_real[5],
                tau      = par_real[6],
                eta.null = par_real[7]
              ),
              dyn_rho_max = FALSE
            )
        },
        error = function(e) rep(NA_real_, nrow(joined))
      )

      joined %>%
        mutate(swe_mod = swe_mod) %>%
        drop_na()
    })

    bind_rows(block_results)
  }

  dff <- bind_rows(station_results)

  if (nrow(dff) == 0) {
    cat("No valid model output. Returning large penalty.\n")
    return(1e12)
  }

  score <- combined_rmse(dff)

  if (!is.finite(score)) {
    cat("Score is not finite. Returning large penalty.\n")
    return(1e12)
  }

  rmse_swe <- rmse(dff$swe_obs, dff$swe_mod)

  rho_obs <- ifelse(dff$hs > 1e-6, dff$swe_obs / dff$hs, NA_real_)
  rho_mod <- ifelse(dff$hs > 1e-6, dff$swe_mod / dff$hs, NA_real_)
  rmse_rho <- rmse(rho_obs, rho_mod)

  bias_swe <- mean(dff$swe_mod - dff$swe_obs, na.rm = TRUE)

  cat(
    "| bias_swe =", round(bias_swe, 4),
    "| rmse_swe =", round(rmse_swe, 4),
    "| rmse_density =", round(rmse_rho, 4),
    "| final_score =", round(score, 4),
    "\n"
  )

  score
}

##############################################################################
# START VALUES AND SCALING
##############################################################################

par_delta <- c(
  rho.max  = 396.9521,
  rho.null = 78.88324,
  c.ov     = 0.0004986025,
  k.ov     = 0.227786,
  k        = 0.0290256,
  tau      = 0.02489556,
  eta.null = 8792253
)

par_scale <- c(1000, 1000, 0.001, 1, 0.1, 0.1, 1e7)

par_start <- par_delta / par_scale

print(par_start)

##############################################################################
# PARALLEL SETUP
##############################################################################

nc <- parallel::detectCores(logical = TRUE) - 1
nc <- max(1, nc)

cl <- parallel::makeCluster(nc)
doParallel::registerDoParallel(cl)

##############################################################################
# TEST OBJECTIVE FUNCTION ON START VALUES
##############################################################################

test_score <- minimize_score(
  par   = par_start,
  data  = d_obs_fit_tibble,
  scale = par_scale,
  verbose = FALSE
)

cat("Initial score =", test_score, "\n")

##############################################################################
# OPTIMIZATION
##############################################################################

opt <- optimx(
  par     = par_start,
  fn      = minimize_score,
  data    = d_obs_fit_tibble,
  scale   = par_scale,
  verbose = FALSE,
  method  = "Nelder-Mead",
  control = list(trace = 1, maxit = 500)
)

##############################################################################
# SAVE RESULTS
##############################################################################

saveRDS(opt, file = "opt_results_combined_rmse_nelder.rds")

##############################################################################
# STOP CLUSTER
##############################################################################

parallel::stopCluster(cl)

File: calibration/archieve/dsnow_parameter_optimization.py
##############################################################################
# DELTASNOW PARAMETER OPTIMIZATION FROM SMET FILES
##############################################################################

import sys
import contextlib
import io
import numpy as np
import pandas as pd
from pathlib import Path
from scipy.optimize import minimize
from joblib import Parallel, delayed

sys.path.insert(0, "/Users/jakobwerkgarner/code/mt_dsnow/snow_to_swe_master")
from main import SnowToSwe

##############################################################################
# PATHS AND SETTINGS
##############################################################################

INPUT_DIR = Path("/Users/jakobwerkgarner/code/mt_dsnow/par_sens/SNOWPACK_data")

SEASON_START_MONTH = 8
MIN_WINTER_LENGTH = 200
EPS = 1e-6
N_JOBS = -1

##############################################################################
# OBJECTIVE WEIGHTS  (should sum to 1)
##############################################################################

WEIGHT_SWE_NRMSE = 0.5   # normalized RMSE of SWE
WEIGHT_RHO_NRMSE = 0.5   # normalized RMSE of bulk density
WEIGHT_SWE_NBIAS = 0.0   # normalized absolute bias of SWE

##############################################################################
# VARIABLES
##############################################################################

HS_COL = "HS_meas"
SWE_COL = "SWE"


import warnings
warnings.filterwarnings("ignore", category=FutureWarning)

##############################################################################
# SMET FUNCTIONS
##############################################################################

def read_smet(file_path):
    header = []
    fields = None
    data_rows = []
    data_start = False

    with open(file_path, "r") as f:
        for line in f:
            line = line.rstrip()

            if line.startswith("fields"):
                fields = line.split("=", 1)[1].strip().split()

            elif line.strip() == "[DATA]":
                data_start = True
                continue

            elif not data_start:
                header.append(line)

            else:
                if line:
                    data_rows.append(line.split())

    df = pd.DataFrame(data_rows, columns=fields)
    df["timestamp"] = pd.to_datetime(df["timestamp"])

    return header, df


def extract_station_id(header):
    for line in header:
        if line.startswith("station_id"):
            return line.split("=", 1)[1].strip()
    raise ValueError("station_id not found in SMET header")


##############################################################################
# GENERAL HELPER FUNCTIONS
##############################################################################

def rmse(obs, mod):
    obs = np.asarray(obs, dtype=float)
    mod = np.asarray(mod, dtype=float)

    ok = np.isfinite(obs) & np.isfinite(mod)
    if not np.any(ok):
        return np.nan

    return np.sqrt(np.mean((obs[ok] - mod[ok]) ** 2))


def combined_score(df, eps=1e-6):
    swe_obs = df["swe_obs"].to_numpy()
    swe_mod = df["swe_mod"].to_numpy()
    hs      = df["hs"].to_numpy()

    # Bulk density (only where snow is present)
    rho_obs = np.where(hs > eps, swe_obs / hs, np.nan)
    rho_mod = np.where(hs > eps, swe_mod / hs, np.nan)

    rmse_swe = rmse(swe_obs, swe_mod)
    rmse_rho = rmse(rho_obs, rho_mod)

    mean_swe = np.nanmean(swe_obs)
    mean_rho = np.nanmean(rho_obs)

    ok = np.isfinite(swe_obs) & np.isfinite(swe_mod)
    bias_swe = np.mean(swe_mod[ok] - swe_obs[ok]) if np.any(ok) else np.nan

    nrmse_swe = rmse_swe / mean_swe  if (np.isfinite(rmse_swe) and mean_swe > 0) else np.inf
    nrmse_rho = rmse_rho / mean_rho  if (np.isfinite(rmse_rho) and mean_rho > 0) else np.inf
    nbias_swe = abs(bias_swe) / mean_swe if (np.isfinite(bias_swe) and mean_swe > 0) else np.inf

    score = (
        WEIGHT_SWE_NRMSE * nrmse_swe +
        WEIGHT_RHO_NRMSE * nrmse_rho +
        WEIGHT_SWE_NBIAS * nbias_swe
    )
    return score if np.isfinite(score) else np.inf


def is_valid_winter(df):
    if len(df) < MIN_WINTER_LENGTH:
        return False

    if len(df) < 365:
        first_hs = float(df["hs"].iloc[0])
        last_hs = float(df["hs"].iloc[-1])
        if first_hs > 0.05 or last_hs > 0.05:
            return False

    return True


##############################################################################
# LOAD ALL SMET FILES
##############################################################################

def load_all_smet_data(input_dir):
    all_data = []

    for smet_file in input_dir.glob("*.smet"):
        print(f"Reading {smet_file.name}")

        header, df = read_smet(smet_file)
        station_id = extract_station_id(header)

        needed = ["timestamp", HS_COL, SWE_COL]
        missing = [col for col in needed if col not in df.columns]
        if missing:
            raise ValueError(f"{smet_file.name} is missing columns: {missing}")

        df = df[needed].copy()

        df[[HS_COL, SWE_COL]] = (
            df[[HS_COL, SWE_COL]]
            .apply(pd.to_numeric, errors="coerce")
            .replace(-999, np.nan)
        )

        df_daily = (
            df.set_index("timestamp")[[HS_COL, SWE_COL]]
            .resample("D")
            .mean()
            .reset_index()
        )

        df_daily["name"] = station_id
        df_daily = df_daily.rename(
            columns={
                "timestamp": "date",
                HS_COL: "hs",
                SWE_COL: "swe_obs"
            }
        )

        all_data.append(df_daily)

    if len(all_data) == 0:
        raise ValueError("No SMET files found.")

    out = pd.concat(all_data, ignore_index=True)
    out["date"] = pd.to_datetime(out["date"])
    out = out.sort_values(["name", "date"]).reset_index(drop=True)

    return out


##############################################################################
# PREPARE FIT DATA
##############################################################################

def prepare_fit_data(df_all, start_month=8):
    fit_parts = []

    for station, df_station in df_all.groupby("name"):
        df_station = df_station.sort_values("date").copy()
        years = np.sort(df_station["date"].dt.year.unique())

        print(f"Processing station: {station}")

        for y in years[:-1]:
            start = pd.Timestamp(year=int(y), month=8, day=1)
            end = pd.Timestamp(year=int(y) + 1, month=7, day=31)

            winter = df_station[
                (df_station["date"] >= start) &
                (df_station["date"] <= end)
            ].copy()

            if winter.empty:
                continue

            if not is_valid_winter(winter):
                print(f"  skipping winter {y}/{y+1}")
                continue

            winter["block"] = int(y)
            fit_parts.append(winter)

    if len(fit_parts) == 0:
        raise ValueError("No valid winters found for fitting.")

    fit_data = pd.concat(fit_parts, ignore_index=True)
    return fit_data


##############################################################################
# RUN DELTASNOW FOR ONE STATION/BLOCK
##############################################################################

def run_one_block(df_block, par_real):
    df_block = df_block.sort_values("date").copy()

    full_dates = pd.date_range(
        start=df_block["date"].min(),
        end=df_block["date"].max(),
        freq="D"
    )

    joined = (
        df_block[["date", "hs", "swe_obs"]]
        .set_index("date")
        .reindex(full_dates)
        .rename_axis("date")
        .reset_index()
    )

    # SnowToSwe requires no NaNs and first value == 0
    hs_series = (
        joined["hs"]
        .interpolate(method="linear", limit_direction="both")
        .clip(lower=0.0)
    )
    if hs_series.isna().any():
        return pd.DataFrame()  # skip block if interpolation still leaves NaN
    hs_arr = hs_series.to_numpy(dtype=float)
    hs_arr[0] = 0.0

    try:
        with contextlib.redirect_stdout(io.StringIO()):
            model = SnowToSwe(
                rho_max=par_real[0],
                rho_null=par_real[1],
                c_ov=par_real[2],
                k_ov=par_real[3],
                k=par_real[4],
                tau=par_real[5],
                eta_null=par_real[6],
            )
        swe_mod = model.convert_list(hs_arr.tolist(), timestep=24)
        joined["swe_mod"] = np.asarray(swe_mod, dtype=float)
    except Exception:
        joined["swe_mod"] = np.nan

    joined = joined.dropna(subset=["hs", "swe_obs", "swe_mod"]).copy()
    return joined


##############################################################################
# OBJECTIVE FUNCTION
##############################################################################

def objective(par_scaled, data, par_scale, verbose=False):
    par_real = np.asarray(par_scaled) * np.asarray(par_scale)

    groups = [g for _, g in data.groupby(["name", "block"], sort=False)]

    results = Parallel(n_jobs=N_JOBS)(
        delayed(run_one_block)(g, par_real) for g in groups
    )

    results = [r for r in results if r is not None and not r.empty]

    if len(results) == 0:
        return 1e12

    dff = pd.concat(results, ignore_index=True)

    score = combined_score(dff, eps=EPS)

    if not np.isfinite(score):
        return 1e12

    if verbose:
        rmse_swe = rmse(dff["swe_obs"], dff["swe_mod"])
        rho_obs = np.where(dff["hs"] > EPS, dff["swe_obs"] / dff["hs"], np.nan)
        rho_mod = np.where(dff["hs"] > EPS, dff["swe_mod"] / dff["hs"], np.nan)
        rmse_rho = rmse(rho_obs, rho_mod)
        bias_swe = np.nanmean(dff["swe_mod"] - dff["swe_obs"])

        print(
            f"| bias_swe = {bias_swe:.4f} "
            f"| rmse_swe = {rmse_swe:.4f} "
            f"| rmse_density = {rmse_rho:.4f} "
            f"| final_score = {score:.4f}"
        )

    return float(score)


##############################################################################
# START VALUES
##############################################################################

par_delta = np.array([
    401,
    81,
    5.1e-4,
    0.38,
    0.03,
    0.024,
    8.5e6,      
    ], dtype=float)

par_scale = np.array([
    1000,
    1000,
    0.001,
    1,
    0.1,
    0.1,
    1e7
], dtype=float)

##############################################################################
# SCALE START VALUES
##############################################################################

par_start = par_delta / par_scale

print("\nScaled start parameters:")
print(par_start)

print("\nUnscaled start parameters:")
print(par_start * par_scale)

##############################################################################
# CALLBACK TO PRINT OPTIMIZATION PROGRESS
##############################################################################

iteration = 0

def print_progress(xk):
    global iteration
    iteration += 1

    par_real = xk * par_scale
    score = objective(xk, fit_data, par_scale, verbose=False)

    print(
        f"\nITER {iteration}"
        f"\nparams = {np.round(par_real, 6)}"
        f"\nscore  = {score:.6f}"
    )

##############################################################################
# MAIN
##############################################################################

df_all = load_all_smet_data(INPUT_DIR)
fit_data = prepare_fit_data(df_all, start_month=SEASON_START_MONTH)

print("\nTesting objective function ...")
test_score = objective(par_start, fit_data, par_scale, verbose=True)
print("Initial score:", test_score)

print("\nStarting optimization ...")

res = minimize(
    fun=objective,
    x0=par_start,
    args=(fit_data, par_scale, False),
    method="Nelder-Mead",
    callback=print_progress,
    options={
        "maxiter": 2000,
        "xatol": 1e-6,
        "fatol": 1e-6,
    }
)

##############################################################################
# RESULTS
##############################################################################

best_par_scaled = res.x
best_par_real = best_par_scaled * par_scale

param_names = ["rho_max", "rho_null", "c_ov", "k_ov", "k", "tau", "eta_null"]
best_params = pd.Series(best_par_real, index=param_names, name="best_value")

print("\nBest parameters:")
print(best_params)

print("\nFinal score:")
print(res.fun)

print("\nOptimizer message:")
print(res.message)